Allow _FilterFloat to pass None values.

Ouwen opened the pull request:
[Tensorflow 1.3: tf.constant with dtype=[float32, float64, float16] may have inconsistent behavior. #13827](https://github.com/tensorflow/tensorflow/issues/13827)

 - This fix allows _FilterFloat to allow None values to pass.
 - We add `tf.float16` into the `_TF_TO_IS_OK` dictionary.
 - We add tests for `convert_to_tensor` and `tensor_util`.

Ouwen committed:
Allow _FilterFloat to pass None values.

[Tensorflow 1.3: tf.constant with dtype=[float32, float64, float16] may have inconsistent behavior. #13827](https://github.com/tensorflow/tensorflow/issues/13827)

 - This fix allows _FilterFloat to allow None values to pass.
 - We add `tf.float16` into the `_TF_TO_IS_OK` dictionary.
 - We add tests for `convert_to_tensor` and `tensor_util`.

tensorflow-jenkins commented:
Can one of the admins verify this patch?

vrv commented on tensorflow/python/framework/tensor_util.py:
```diff
@@ -252,7 +252,7 @@ def _FilterInt(v):
   if isinstance(v, (list, tuple)):
     return _FirstNotNone([_FilterFloat(x) for x in v])
-  return None if isinstance(v, compat.real_types) else _NotNone(v)
+  return None if isinstance(v, compat.real_types) or v is None else _NotNone(v)
```
I don't understand this change: _NotNone(v) handles None in a specific way that seems intentional?  Can you clarify why you want to do this?

vrv replied:
Looking at the original PR, I am going to defer to someone more of an expert on convert_to_tensor().  I don't know what it means to have a None value in an array constant.  It seems like this is targeting some kind of numpy parity but I don't know what numpy wants to do in the first place.

mrry replied:
I'd defer further to @josh11b, since he wrote the original code that intentionally raises an exception (albeit perhaps not with the message intended) when `None` is present in a `tf.float32` or `tf.float64` constant. It looks like `tf.float16` slipped through the cracks when it was added, and we defer to NumPy semantics.

(My 2 cents: NumPy parity seems worth trying to achieve here, since I don't think it comes at a cost to backwards compatibility. However, I'm not sure what all the consequences of the change would be.)

vrv replied:
Thanks @mrry, I'll defer to @josh11b then.  I don't know what it means to have a Tensor with some known values and some None values.

MarkDaoust replied:
> Can you clarify why you want to do this?

The goal is better compatibility with numpy (this is described in the linked bug).

You can create a numpy float array from a list containing `None` and the `None` gets replaced by a `NaN`.

So the goal here is to accept `None` in the input data. 

Maybe this makes the call to `_NotNone()` redundant since, `_FilterFloat` returns `None` for success and the failing value on failure. So the purpose of `_NotNone` is to wrap `None` values so that failure can be reported for a `None` value.

I am no expert in this code either, this is just what I've understood from being involved in the bug.

Ouwen replied:
> So the purpose of _NotNone is to wrap None values so that failure can be reported for a None value.

Right, so if the wrong data type is placed into a `tf.float32` tensor. The correct error message is returned.
```
sess = tf.Session()
print(sess.run(tf.constant([1, 2, 3, 4, 5, 6, 'foo'], dtype=tf.float32)))
# TypeError: Expected float32, got 'string' of type 'str' instead.
```

numpy floats allow `nan` values and numpy converts `None` to `nan`; however, the current `make_tensor_proto` and `tf.constant` do not convert `None` to `nan` for float types.

mrry replied:
Right. I'm not sure if exact compatibility with NumPy was the goal with @josh11b added the `_NotNone()` path. It's possible that there was an overriding reason to exclude `None` values, which is why I'd like Josh to cast the deciding vote :).

alextp replied:
I do not think silently converting None to NaN would be a good idea in general. At least from my personal experience the check to prevent Nones from being silently converted to tensors has saved me many hours of debugging weird issues.

Ouwen replied:
@alextp, thanks for giving this a look over!

Currently there exists inconsistent behavior since `numpy.array` will convert `None` into `nan` for float types but, `convert_to_tensor` does not. This seems to break the promised `convert_to_tensor` abstraction referenced from issue #13827.

>Since the `convert_to_tensor` abstraction states the [following are equivalent](https://github.com/tensorflow/tensorflow/blob/r1.3/tensorflow/python/framework/ops.py#L575-L578)
> ```
> value_1 = my_func(tf.constant([[1.0, 2.0], [3.0, 4.0]]))
> value_2 = my_func([[1.0, 2.0], [3.0, 4.0]])
> value_3 = my_func(np.array([[1.0, 2.0], [3.0, 4.0]], dtype=np.float32))
> ```
>I believe that the below should be equivalent as well
> ```
> value_a = my_func(tf.constant([[1.0, 2.0], [3.0, None]], dtype=tf.float32))
> value_b = my_func(np.array([[1.0, 2.0], [3.0, None]], dtype=np.float32))
> ```

From this inconsistency in `convert_to_tensor`, I believe there are a couple options.
1. We document that the abstraction does not hold for floats.
2. We remove the `convert_to_tensor` abstraction
3. We update `NumPy` to not convert `None` to `nan` for float types
4. We silently convert `None` to `nan` specifically for float types

Options 1 and 4 seem reasonable to me, but in my opinion option 4 is a good choice given that `convert_to_tensor` already silently converts and that it follows the semantics `NumPy` has set forth.

Please let me know what you think. Happy to make any needed changes!

alextp commented:
I think we should do option (1). There are already a few places in which
convert_to_tensor disagrees with numpy (about what dtype to use and about
what to do with strings containing nul bytes, for example).

On Sun, Oct 29, 2017 at 9:58 PM, Ouwen Huang <notifications@github.com>
wrote:

> *@Ouwen* commented on this pull request.
> ------------------------------
>
> In tensorflow/python/framework/tensor_util.py
> <https://github.com/tensorflow/tensorflow/pull/13834#discussion_r147617934>
> :
>
> > @@ -252,7 +252,7 @@ def _FilterInt(v):
>  def _FilterFloat(v):
>    if isinstance(v, (list, tuple)):
>      return _FirstNotNone([_FilterFloat(x) for x in v])
> -  return None if isinstance(v, compat.real_types) else _NotNone(v)
> +  return None if isinstance(v, compat.real_types) or v is None else _NotNone(v)
>
> @alextp <https://github.com/alextp>, thanks for giving this a look over!
>
> Currently there exists inconsistent behavior since numpy.array will
> convert None into nan for float types but, convert_to_tensor does not.
> This seems to break the promised convert_to_tensor abstraction referenced
> from issue #13827 <https://github.com/tensorflow/tensorflow/issues/13827>.
>
> Since the convert_to_tensor abstraction states the following are
> equivalent
> <https://github.com/tensorflow/tensorflow/blob/r1.3/tensorflow/python/framework/ops.py#L575-L578>
>
> value_1 = my_func(tf.constant([[1.0, 2.0], [3.0, 4.0]]))
> value_2 = my_func([[1.0, 2.0], [3.0, 4.0]])
> value_3 = my_func(np.array([[1.0, 2.0], [3.0, 4.0]], dtype=np.float32))
>
> I believe that the below should be equivalent as well
>
> value_a = my_func(tf.constant([[1.0, 2.0], [3.0, None]], dtype=tf.float32))
> value_b = my_func(np.array([[1.0, 2.0], [3.0, None]], dtype=np.float32))
>
> From this inconsistency in convert_to_tensor, I believe there are a
> couple options.
>
>    1. We document that the abstraction does not hold for floats.
>    2. We remove the convert_to_tensor abstraction
>    3. We update NumPy to not convert None to nan for float types
>    4. We silently convert None to nan specifically for float types
>
> Options 1 and 4 seem reasonable to me, but in my opinion option 4 is a
> good choice given that convert_to_tensor already silently converts and
> that it follows the semantics NumPy has set forth.
>
> Please let me know what you think. Happy to make any needed changes!
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/tensorflow/tensorflow/pull/13834#discussion_r147617934>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAATxav419Oc9w5QbrSLn20p9YCtNi07ks5sxVdmgaJpZM4P_iGB>
> .
>



-- 
 - Alex

Ouwen commented:
@alextp, you have a good point with NumPy  having different behavior when handling `null` types for strings. If `convert_to_tensor` does not fall back on NumPy semantics, perhaps it is better to allow users to explicitly decide the method of conversion.

With that, I can update the PR so that we add `tf.float16` into the `_TF_TO_IS_OK` dictionary for consistent assertion. Additionally, I will make a note that tensors with `None` values can have differing behavior from NumPy in the documentation.

If this sounds good to you guys @alextp  @mrry @MarkDaoust @vrv, I can go ahead and make that change.

Ouwen commented:
@alextp @josh11b @mrry @MarkDaoust @vrv, any updates? Just want to make sure this is the plan moving forward.

An alternative would be to potentially default to numpy semantics for all tensors that are made.

https://github.com/tensorflow/tensorflow/blob/408fd454d7d2a16269576ea12bcd516e25a6b0c5/tensorflow/python/framework/tensor_util.py#L375-L377

So the following lines would be flipped to attempt a numpy conversion, and if that fails we can fall back to the compatibility error.

This would mean `None` would be converted to `nan` for floats, and that `None` would be converted to `'None'` for strings.

MarkDaoust commented:
Hi Ouwen,

Thanks for the ping.

>  I can update the PR so that we add tf.float16 into the _TF_TO_IS_OK dictionary for 
> consistent assertion. Additionally, I will make a note that tensors with None values can have
> differing behavior from NumPy in the documentation.

Both vrv and mrry suggested josh11b decide, and josh11b sent alextp. So I think the `None` issue is settled, and this is the right direction.

I think everyone agrees that float16 should follow the same rules as the other floats, and that the discrepancy between tf and numpy should be documented so the intent is clear (you could also note the strings with null bytes difference that alextp mentioned).

Thanks again.

Ouwen committed:
Added `tf.float16` into the _TF_TO_IS_OK dictionary

[Tensorflow 1.3: tf.constant with dtype=[float32, float64, float16] may have inconsistent behavior. #13827](https://github.com/tensorflow/tensorflow/issues/13827)

This fix adds `tf.float16` into the `_TF_TO_IS_OK` dictionary for consistent error checking.
A note is added to `convert_to_tensor` on how `None` values are handled.

Ouwen commented:
Appreciate the clarification @MarkDaoust! I've put up the PR change

alextp commented:
Thanks for the contribution!

alextp commented:
Jenkins, test this please.
